### PR TITLE
Card: Variant attribute introduced with elevated, flat and outlined as options

### LINKED ISSUE #1998: [Enhancement] Card Outline
## Describe the enhancement
Right now cards only support fill-colors but we would like to have support for outline cards too.

## Describe the solution you'd like
Add Outline as a 3th. visual variant for the existing Card component. (Default, Flat, **Outline**)
Outline should be similar to the existing "Flat" variant in regards to guidelines (not used for clickable cards)
The outline is available in our sketch library but just not available for our cards right now.
The relevant color right now is "grey"/"medium" (#D1D1D1) 

**SPEC:**
Outline stroke 1px - "grey"/"medium" (#D1D1D1)
Background: none

## Have you considered any alternatives?
No

## Are there any additional context?
Here is just one example

Skærmbillede 2022-01-18 kl  09 06 32

**Link to Zeplin:**
https://zpl.io/8Ejm3Ww

<hr />

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Refinement

- [ ] Request that the issue is [UX refined](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#ux-refinement); do not proceed until this is done.
- [ ] Request that the issue is [tech refined](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#tech-refinement); do not proceed until this is done.

### Implementation 
The contributor who wants to implement this issue should: 

- [ ] Make sure you have read: "[Before you get coding](https://github.com/kirbydesign/designsystem/blob/master/.github/CONTRIBUTING.md/#before-you-get-coding)".
- [ ] Signal to others you are working on the issue by assigning yourself.
- [ ] Create a branch from the [master branch](https://github.com/kirbydesign/designsystem/tree/master) following our [branch naming convention](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Branch). 
- [ ] Publish a WIP implementation to Github as a draft PR and ask for feedback. 
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] Update the [cookbook](https://cookbook.kirby.design) with examples and showcases.

### Review
Once the issue has been implemented and is ready for review:

- [ ] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review). 
- [ ] Create a pull-request. If you created a draft PR during implementation you can just mark that as "ready for review".

### PR DESCRIPTION
**Pull request by Fuzzy3**
## Which issue does this PR close?

This PR closes #1998

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/develop/.github/CONTRIBUTING.md#the-process-of-contributing) correctly.

### Reminders
- [x] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [x] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [ ] Determine if your changes are a fix, feature or breaking-change, and add the matching label to your PR. If it is tooling, dependency updates or similar, add ignore-for-release.
- [x] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [x] Request that the changes are code-reviewed 
- [x] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/develop/.github/CONTRIBUTING.md#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be merged to `develop` by Team Kirby.

### CODE REVIEW

File: libs/designsystem/card/src/card.component.ts [hunk 1]
```diff
@@ -45,9 +50,27 @@ export class CardComponent implements OnInit, OnDestroy {
     this.sizesSortedByBreakpoint = this.sortSizesByBreakpoint(value);
   }
 
-  @HostBinding('class.flat')
+  /**
+   * @deprecated Using flat is no longer recomended as it is now part of the "variant" option. Will be removed in version 10.0.
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/card/src/card.component.ts]:
```suggestion
   * @deprecated Setting the flat property is no longer recommended, and can now be set with variant="flat" instead. The 'flat' property will be removed in Kirby v10.
```

Comment (RasmusKjeldgaard) [reply libs/designsystem/card/src/card.component.ts]: Just a small suggestion to give a little more context and some specific migration advice 😊

File: libs/designsystem/card/src/card.component.ts [hunk 2]
```diff
@@ -9,12 +9,17 @@ import {
 } from '@angular/core';
 import { ResizeObserverService } from '@kirbydesign/designsystem/shared';
 
+const KIRBY_CARD_FLAT_PROPERTY_DEPRECATION_WARNING =
+  'Deprecation warning: flat property will be removed from version 10.0. Use variant="flat" instead';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/card/src/card.component.ts]: As this will show up in the browsers console, it is a good idea to add some more context to the message so consumers know where to fix it when they see the message:
```suggestion
  'Deprecation warning: Setting the 'flat' property of kirby-card is deprecated and will be removed from Kirby v10. Use variant="flat" instead for identical behavior.';
```

File: libs/designsystem/card/src/card.component.spec.ts [hunk 3]
```diff
@@ -2,12 +2,16 @@ import { TestHelper } from '@kirbydesign/designsystem/testing';
 import { createHostFactory, SpectatorHost } from '@ngneat/spectator';
 
 import { CardComponent } from './card.component';
+import { CardHeaderComponent } from './public_api';
+
+const OUTLINE_COLOR = 'rgb(209, 209, 209)';
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/card/src/card.component.spec.ts]: You can use the `DesignTokenHelper` to retreive the value via its name instead of hard-coding the value. 
```suggestion
const getColor = DesignTokenHelper.getColor;
```

Comment (RasmusKjeldgaard) [reply libs/designsystem/card/src/card.component.spec.ts]: And then use further down, something like this:
`'border-block-end-color': getColor('medium')`

File: apps/cookbook/src/app/examples/card-example/examples/outline.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { Component } from '@angular/core';
+
+const config = {
+  selector: 'cookbook-card-example-outline',
+  template: `<kirby-card [hasPadding]="true" variant="outlined">
+    <strong>This is an outlined card</strong>
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/outline.ts]: Consider keeping naming in sync with UX:
```suggestion
    <strong>This is an outline card</strong>
```

File: apps/cookbook/src/app/examples/card-example/examples/outline.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { Component } from '@angular/core';
+
+const config = {
+  selector: 'cookbook-card-example-outline',
+  template: `<kirby-card [hasPadding]="true" variant="outlined">
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/outline.ts]: Consider keeping naming in sync with UX:
```suggestion
  template: `<kirby-card [hasPadding]="true" variant="outline">
```

File: apps/cookbook/src/app/examples/card-example/examples/outline.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    <strong>This is an outlined card</strong>
+    <p>This card has no background and does not react on theming</p>
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/outline.ts]:
```suggestion
    <p>This card has no background and does not react on theming.</p>
```

File: apps/cookbook/src/app/examples/card-example/examples/outline.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    <p>In general the look and feel of the default card is preferred in most use cases as it elevates the card a bit from the background,
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/outline.ts]: Consider describing the relevant use cases instead of what it's not 😊  E.g. "The outline variant of the card should be used for ..." Maybe align with @henrikvoetmand 🤞🏻

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html [hunk 8]
```diff
@@ -105,14 +105,24 @@ <h2>Background image</h2>
...
+  by setting the variant to "flat", but be aware that this option should not be used with a card
   that can be interacted with (clickable, disclosure) as it lacks proper affordance.
 </p>
 <cookbook-example-viewer [html]="flatExample.template">
   <cookbook-card-example-flat #flatExample></cookbook-card-example-flat>
 </cookbook-example-viewer>
 
+<h2>Outlined</h2>
+<p>
+  An outlined variant of a card is shown only with a border around its edges and have no background.
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html]: Consider keeping naming in sync with UX:
```suggestion
  An outline variant of a card is shown only with a border around its edges and have no background.
```

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.ts [hunk 9]
```diff
@@ -31,10 +31,18 @@ export class CardShowcaseComponent {
       defaultValue: '',
       type: ['string'],
     },
+    {
+      name: 'variant',
+      description:
+        'Sets the variant of a card. The recommended is "elevated" which is also the default',
+      defaultValue: 'elevated',
+      type: ['elevated', 'flat', 'outlined'],
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.ts:39]: Consider keeping naming in sync with UX:
```suggestion
      type: ['elevated', 'flat', 'outline'],
```

File: libs/designsystem/card/src/card-as-button/card-as-button.directive.ts [hunk 10]
```diff
@@ -15,7 +15,7 @@ export class CardAsButtonDirective implements OnInit {
 
   ngOnInit(): void {
     if (this.card) {
-      this.card.flat = false;
+      this.card.variant = 'elevated';
```
Comment (jakobe) [inline libs/designsystem/card/src/card-as-button/card-as-button.directive.ts]: I would think this is redundant as the card has `variant='elevated'` by default 🤔 
```suggestion
```

File: libs/designsystem/card/src/card.component.scss [hunk 11]
```diff
@@ -57,6 +57,24 @@
...
+
+    &:has(kirby-card-header.flagged) {
+      outline: none;
+
+      .content-wrapper {
+        border-block-end: 1px solid utils.get-color('medium');
+        border-inline-start: 1px solid utils.get-color('medium');
+        border-inline-end: 1px solid utils.get-color('medium');
+        border-bottom-left-radius: 16px;
+        border-bottom-right-radius: 16px;
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.scss]: Should use same value/variable as the overall card border radius:
```suggestion
        border-bottom-left-radius: utils.$border-radius;
        border-bottom-right-radius: utils.$border-radius;
```

File: libs/designsystem/card/src/card.component.scss [hunk 12]
```diff
@@ -57,6 +57,24 @@
     box-shadow: none;
   }
 
+  &.outlined {
+    background-color: transparent;
+    outline: 1px solid utils.get-color('medium');
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.scss]: Is there a reason we use `outline` instead of `border`? 🤔

File: libs/designsystem/card/src/card.component.spec.ts [hunk 13]
```diff
@@ -2,12 +2,16 @@ import { TestHelper } from '@kirbydesign/designsystem/testing';
 import { createHostFactory, SpectatorHost } from '@ngneat/spectator';
 
 import { CardComponent } from './card.component';
+import { CardHeaderComponent } from './public_api';
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.spec.ts]: Consider aligning import with line above:
```suggestion
import { CardHeaderComponent } from './card-header/card-header.component';
```

File: libs/designsystem/slide/src/slides.component.spec.ts [hunk 14]
```diff
@@ -158,6 +158,7 @@ describe('SlidesComponent', () => {
   });
 
   it('should extend the default slides options with the provided slides options', () => {
+    spectator = createHost(hostTemplate, props);
```
Comment (jakobe) [inline libs/designsystem/slide/src/slides.component.spec.ts:161]: Is this related to the issue/PR at hand? 🤔

File: libs/designsystem/card/src/card.component.ts [hunk 15]
```diff
@@ -45,9 +50,27 @@ export class CardComponent implements OnInit, OnDestroy {
...
+    this._flat = isFlat;
+    console.warn(KIRBY_CARD_FLAT_PROPERTY_DEPRECATION_WARNING);
+  }
+
+  get flat(): boolean {
+    return this._flat;
+  }
+
+  @Input()
+  variant: 'elevated' | 'flat' | 'outlined' = 'elevated';
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.ts:67]: Consider keeping naming in sync with UX:
```suggestion
  variant: 'elevated' | 'flat' | 'outline' = 'elevated';
```

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html (hunk 8 above)
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html]:
```suggestion
  An outlined variant of a card is shown only with a border around it's edges and have no background.
```

**Review by jakobe — COMMENTED** [review]
Great work 👏🏻

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.ts [hunk 16]
```diff
@@ -31,10 +31,18 @@ export class CardShowcaseComponent {
...
+    },
     {
       name: 'flat',
-      description: '(Optional) Remove elevation on the card. Default elevation = 2, flat = 0.',
-      defaultValue: '',
+      description:
+        'Deprecation warning: use "variant" instead. \n(Optional) Remove elevation on the card. Default elevation = 2, flat = 0.',
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.ts]: Should we consider removing the `flat` prop from the API docs in this PR, as we don't wan't to promote it's use anyhoo? 🤔

File: libs/designsystem/card/src/card.component.spec.ts [hunk 17]
```diff
@@ -45,36 +49,75 @@ describe('CardComponent', () => {
...
 
-    it('should create nested Ionic component with default Ionic mode', async () => {
-      const document = spectator.query<HTMLElement>('html', { root: true });
-      const ionicGlobalMode = document.getAttribute('mode');
-      const ionBadgeElement = spectator.query<HTMLElement>('ion-badge');
-      expect(ionBadgeElement).toBeTruthy();
-      await TestHelper.whenReady(ionBadgeElement);
-      expect(ionBadgeElement).toHaveClass('ios', ionicGlobalMode);
+        'border-bottom-left-radius': '16px',
+        'border-bottom-right-radius': '16px',
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/card/src/card.component.spec.ts]: Similar to the color you can use `DesignTokenHelper.size()` here instead of pixel values.

File: libs/designsystem/card/src/card.component.spec.ts [hunk 18]
```diff
@@ -45,36 +49,75 @@ describe('CardComponent', () => {
     });
   });
 
-  describe('with mode attribute', () => {
+  describe('with flat variant attribute', () => {
+    let cardElement: HTMLElement;
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/card/src/card.component.spec.ts]: Consider declaring this top level like we do with spectator to avoid repitition 🔁 😄

File: libs/designsystem/card/src/card.component.spec.ts [hunk 19]
```diff
@@ -45,36 +49,75 @@ describe('CardComponent', () => {
...
+
     beforeEach(() => {
-      spectator = createHost('<kirby-card mode="flat"></kirby-card>');
+      spectator = createHost('<kirby-card variant="flat"></kirby-card>');
+      cardElement = spectator.queryHost('kirby-card');
```
Comment (RasmusKjeldgaard) [inline libs/designsystem/card/src/card.component.spec.ts]: Consider using the spectator element property for similar functionality
```suggestion
      cardElement = spectator.element;
```

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html [hunk 20]
```diff
@@ -115,9 +115,10 @@ <h2>Flat</h2>
 
 <h2>Outlined</h2>
 <p>
-  An outlined variant of a card is shown only with a border around its edges and have no background.
-  Like the flat card, it is also without any shadow around its edges and the card itself is not
-  intended to be used for interactions (clickable, disclosure) as this also lacks proper affordance.
+  The outlined card variant is shown with a border around its edges and have no background. Like the
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html]:
```suggestion
  The outlined card variant is shown with a border around it's edges and have no background. Like the
```

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html [hunk 21]
```diff
@@ -115,9 +115,10 @@ <h2>Flat</h2>
...
+  flat card, it is without any shadow around its edges and the card itself is not intended to be
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html]: Consider simplifying:
```suggestion
  flat card it is not elevated as the card itself is not intended to be
```

File: libs/designsystem/card/src/card.component.scss [hunk 22]
```diff
@@ -59,18 +60,18 @@
 
   &.outlined {
     background-color: transparent;
-    outline: 1px solid utils.get-color('medium');
+    border: 1px solid utils.get-color('medium');
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.scss]: I guess we should allow a custom border in outlined cards as well? 🤔 
```suggestion
    border: border: var(--kirby-card-border, 1px solid utils.get-color('medium'));
```

File: libs/designsystem/card/src/card.component.scss [hunk 23]
```diff
@@ -57,6 +58,24 @@
...
+    box-shadow: none;
+
+    &:has(kirby-card-header.flagged) {
+      border: none;
+
+      .content-wrapper {
+        border-block-end: 1px solid utils.get-color('medium');
+        border-inline-start: 1px solid utils.get-color('medium');
+        border-inline-end: 1px solid utils.get-color('medium');
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.scss]: I guess we should allow a custom border in outlined cards as well? 🤔 
```suggestion
        border-block-end: var(--kirby-card-border, 1px solid utils.get-color('medium'));
        border-inline-start: var(--kirby-card-border, 1px solid utils.get-color('medium'));
        border-inline-end: var(--kirby-card-border, 1px solid utils.get-color('medium'));
```

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.ts [hunk 24]
```diff
@@ -38,16 +38,10 @@ export class CardShowcaseComponent {
...
-      description:
-        'Deprecation warning: use "variant" instead. \n(Optional) Remove elevation on the card. Default elevation = 2, flat = 0.',
-      defaultValue: 'false',
-      type: ['boolean'],
-    },
     {
       name: 'hasPadding',
-      description: '',
+      description:
+        'If set, the card will appear with a white space between the border and the content',
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.ts]: Consider if this should be either "whitespace" or simply "spacing":
```suggestion
        'If set, the card will have spacing between the border and the content',
```

File: apps/cookbook/src/app/examples/card-example/examples/card-example.shared.scss [hunk 25]
```diff
@@ -79,3 +84,27 @@ button {
     transform: scale(0.95);
   }
 }
+
+.variant-card-container {
+  width: 100%;
+  height: 100%;
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/card-example.shared.scss]: Consider controlling width through flex instead:
```suggestion
```

File: apps/cookbook/src/app/examples/card-example/examples/card-example.shared.scss [hunk 26]
```diff
@@ -79,3 +84,27 @@ button {
...
+  display: flex;
+  justify-content: center;
+  flex-wrap: wrap;
+  gap: 16px;
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/card-example.shared.scss]: Replace magic string with defined size:
```suggestion
  gap: utils.size('s');
```

File: apps/cookbook/src/app/examples/card-example/examples/card-example.shared.scss [hunk 27]
```diff
@@ -79,3 +84,27 @@ button {
...
+      max-width: none !important;
+      width: 100%;
+    }
+  }
+
+  kirby-card {
+    max-width: 50%;
+    min-width: 250px;
+    flex: 1;
+  }
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/card-example.shared.scss]: Consider reordering rulesets to avoid `!important` flag and have default styles declared first.
Also use flex to control width of cards:
```suggestion
  kirby-card {
    max-width: 50%;
    min-width: 250px;
    flex: 1;
  }
  
  @container (width < 616px) {
    flex-direction: column;
    align-items: stretch;

    kirby-card {
      max-width: initial;
    }
  }


```

File: apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html [hunk 28]
```diff
@@ -7,21 +7,40 @@
...
 </ul>
 
-<h2>Elevated (default)</h2>
-
 <cookbook-example-viewer [html]="defaultExample.template">
-  <cookbook-card-example-default #defaultExample></cookbook-card-example-default>
+  <cookbook-card-example-variant #defaultExample></cookbook-card-example-variant>
 </cookbook-example-viewer>
 
+<p>The card and its variants supports different use cases:</p>
```
Comment (jakobe) [inline apps/cookbook/src/app/showcase/card-showcase/card-showcase.component.html:31]:
```suggestion
<p>The card and it's variants supports different use cases:</p>
```

File: libs/designsystem/card/src/card.component.scss [hunk 29]
```diff
@@ -46,30 +46,22 @@
...
     &:has(kirby-card-header.flagged) {
       border: none;
 
       .content-wrapper {
-        border-block-end: 1px solid utils.get-color('medium');
-        border-inline-start: 1px solid utils.get-color('medium');
-        border-inline-end: 1px solid utils.get-color('medium');
+        border-block-end: var(--kirby-card-border, 1px solid utils.get-color('medium'));
+        border-inline-start: var(--kirby-card-border, 1px solid utils.get-color('medium'));
+        border-inline-end: var(--kirby-card-border, 1px solid utils.get-color('medium'));
```
Comment (jakobe) [inline libs/designsystem/card/src/card.component.scss]: Consider extracting to Sass variable:
```suggestion
        $border-outline: 1px solid utils.get-color('medium')-,
        border-block-end: var(--kirby-card-border, $border-outline);
        border-inline-start: var(--kirby-card-border, $border-outline);
        border-inline-end: var(--kirby-card-border, $border-outline);
```

**Review by jakobe — COMMENTED** [review]
Looking good 👌🏻 
Just a few minor nitty gritties 🤓

File: apps/cookbook/src/app/examples/card-example/examples/card-example-variant.scss [hunk 30]
```diff
@@ -0,0 +1,29 @@
+@use '@kirbydesign/core/src/scss/utils';
+@use './card-example.shared.scss';
+
+:host {
+  container-type: inline-size;
+  display: flex;
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/card-example-variant.scss]: Guess this should simply be:
```suggestion
  display: block;
```

File: apps/cookbook/src/app/examples/card-example/examples/card-example-variant.scss [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+  flex-wrap: wrap;
+  gap: utils.size('s');
+
+  kirby-card {
+    max-width: 50%;
+    min-width: 250px;
+    flex: 1;
+  }
+
+  @container (width < 616px) {
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/card-example/examples/card-example-variant.scss]: Could we replace this magic number with a calculation? 🤔

**Review by jakobe — APPROVED** [review]
Nice! 👌🏻
